src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -357,6 +357,66 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncodeInternal(string originalString)
+        {
+            fixed (char* pch = originalString)
+            {
+                int safeCharCount = 0;
+                int spaceCharCount = 0;
+                for (int i = 0; i < originalString.Length; i++)
+                {
+                    char c = pch[i];
+                    if (c == ' ')
+                    {
+                        spaceCharCount++;
+                    }
+                    else if (IsUrlSafeChar(c))
+                    {
+                        safeCharCount++;
+                    }
+                }
+
+                if (spaceCharCount == 0 && safeCharCount == originalString.Length)
+                {
+                    // Nothing to expand
+                    return originalString;
+                }
+                
+                int unsafeByteCount = Encoding.UTF8.GetByteCount(pch, originalString.Length) - safeCharCount - spaceCharCount;
+
+                byte[] originalBytes = new byte[safeCharCount + spaceCharCount + unsafeByteCount];
+                byte[] newBytes = new byte[originalBytes.Length + unsafeByteCount * 2];
+
+                fixed (byte* originalBytesPointer = originalBytes)
+                fixed (byte* newBytesPointer = newBytes)
+                {
+                    Encoding.UTF8.GetBytes(pch, originalString.Length, originalBytesPointer, originalBytes.Length);
+
+                    int pos = 0;
+                    for (int i = 0; i < originalBytes.Length; i++)
```

Comment:
This loop should be refactored into separate method and shared between `byte[]` and `string` handling codepaths.